packages/react-query/src/shared/hooks/types.ts

```diff
@@ -160,13 +167,81 @@ export interface UseTRPCMutationOptions<
 > extends UseMutationOptions<TOutput, TError, TInput, TContext>,
     TRPCUseQueryBaseOptions {}
 
-type inferAsyncIterableYield<T> = T extends AsyncIterable<infer U> ? U : T;
 export interface UseTRPCSubscriptionOptions<TOutput, TError> {
   enabled?: boolean;
   onStarted?: () => void;
-  onData: (data: inferAsyncIterableYield<TOutput>) => void;
+  onData?: (data: TOutput) => void;
   onError?: (err: TError) => void;
+  onStateChange?: (state: TRPCConnectionState<TError>) => void;
+}
+
+export interface UseTRPCSubscriptionConnectionIdleResult<_TError> {
+  connectionState: ConnectionIdleState['state'];
+  connectionError: ConnectionIdleState['error'];
+}
+
+export interface UseTRPCSubscriptionConnectionConnectingResult<TError> {
+  connectionState: ConnectionConnectingState<TError>['state'];
+  connectionError: ConnectionConnectingState<TError>['error'];
+}
+
+export interface UseTRPCSubscriptionConnectionPendingResult<_TError> {
+  connectionState: ConnectionPendingState['state'];
+  connectionError: ConnectionPendingState['error'];
+}
+
+export interface UseTRPCSubscriptionConnectionErrorResult<TError> {
+  connectionState: ConnectionErrorState<TError>['state'];
+  connectionError: ConnectionErrorState<TError>['error'];
+}
+
+export type TRPCSubscriptionConnectionState<TError> =
+  | UseTRPCSubscriptionConnectionIdleResult<TError>
+  | UseTRPCSubscriptionConnectionConnectingResult<TError>
+  | UseTRPCSubscriptionConnectionPendingResult<TError>
+  | UseTRPCSubscriptionConnectionErrorResult<TError>;
+
+export interface TRPCSubscriptionBaseResult<TOutput, TError> {
+  status: 'idle' | 'connecting' | 'pending' | 'error';
+  data: undefined | TOutput;
+  error: undefined | TError;
+}
+
+export interface TRPCSubscriptionIdleResult<TOutput> {
+  status: 'idle';
+  data: TOutput;
+  error: undefined;
 }
+
+export interface TRPCSubscriptionConnectingResult<TOutput, TError>
+  extends TRPCSubscriptionBaseResult<TOutput, TError> {
+  status: 'connecting';
+  data: undefined | TOutput;
+  error: undefined;
+}
+
+export interface TRPCSubscriptionPendingResult<TOutput>
+  extends TRPCSubscriptionBaseResult<TOutput, undefined> {
+  status: 'pending';
+  data: TOutput;
+  error: undefined;
+}
+
+export interface TRPCSubscriptionErrorResult<TOutput, TError>
+  extends TRPCSubscriptionBaseResult<TOutput, TError> {
+  status: 'error';
+  data: TOutput | undefined;
+  error: TError;
+}
+
+export type TRPCSubscriptionResult<TOutput, TError> = (
+  | TRPCSubscriptionIdleResult<TOutput>
+  | TRPCSubscriptionConnectingResult<TOutput, TError>
+  | TRPCSubscriptionErrorResult<TOutput, TError>
+  | TRPCSubscriptionPendingResult<TOutput>
+) &
+  TRPCSubscriptionConnectionState<TError>;
```

Comment:
make it a prop instead
```suggestion
```

Reply:
Did we decide not to go ahead with this idea? I'm happy with shallow anyway, it's simpler all round

Reply:
To be honest I'm not clear why we need a separate connectionState and connectionError they seem a subset of the status and error keys already, why not roll them together and force people to handle this in one way?

Reply:
> To be honest I'm not clear why we need a separate connectionState and connectionError they seem a subset of the status and error keys already, why not roll them together and force people to handle this in one way?

- WebSockets
- Error gets ambiguous (I think?)

Reply:
Maybe you're right akshually.

If we wanna introspect websockets we could do a special hook for the wsClient too I guess.... 🤔